### PR TITLE
storage: add storage.sstable.compression_algorithm cluster setting

### PR DESCRIPTION
**Pull request by jbowens**
Introduce a new cluster setting that allows the operator to configure the compression algorithm used when compressing sstable blocks. This allows operators to opt into use of zstd (as opposed to the previous setting of snappy). ZSTD typically achieves better compression ratios than snappy, and operators may find that they can achieve higher node densities through enabling zstd. Future releases may change the default compression algorithm.

In a side-by-side comparison of a 10000-warehouse tpcc import, the zstd cluster achieved a higher import speed of 146 MiB/s versus snappy's 135 MiB/s. The zstd cluster's physical database size was significantly less (~30%).

Screenshot 2024-03-20 at 3 07 13 PM
Screenshot 2024-03-20 at 3 06 49 PM

Informs #105568.
Epic: none
Release note (ops change): Add `storage.sstable.compression_algorithm` cluster setting that configures the compression algorithm to use when compressing sstable blocks.

### CODE REVIEW

File: pkg/storage/pebble.go [hunk 1]
```diff
@@ -132,6 +132,44 @@ var IngestAsFlushable = settings.RegisterBoolSetting(
...
+const (
+	compressionAlgorithmSnappy int64 = 1
+	compressionAlgorithmZstd   int64 = 2
+)
+
+// compressionAlgorithm determines the compression algorithm used to compress
+// data blocks when writing sstables. Users should call getCompressionAlgorithm
+// rather than calling compressionAlgorithm.Get directly.
+var compressionAlgorithm = settings.RegisterEnumSetting(
+	settings.SystemVisible,
```
Comment (itsbilal) [inline pkg/storage/pebble.go:144]: maybe settings.SystemOnly is safer? but also good with this

**Review by itsbilal — APPROVED** [review]
LGTM

File: pkg/storage/pebble.go [hunk 2]
```diff
@@ -132,6 +132,44 @@ var IngestAsFlushable = settings.RegisterBoolSetting(
...
+
+func getCompressionAlgorithm(settings *cluster.Settings) pebble.Compression {
+	switch compressionAlgorithm.Get(&settings.SV) {
+	case compressionAlgorithmSnappy:
+		return pebble.SnappyCompression
+	case compressionAlgorithmZstd:
+		// Pre-24.1 Pebble's implementation of zstd had bugs that could cause
+		// in-memory corruption. We require that the 24.1 version be finalized
+		// before zstd will have any effect.
+		if settings.Version.ActiveVersionOrEmpty(context.TODO()).IsActive(clusterversion.V24_1Start) {
```
Comment (dt) [inline pkg/storage/pebble.go:164]: you do have a ctx in scope in `newPebble` where you call  this if you wanted it.

**Review by dt — APPROVED** [review]

**Review by jbowens — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120784)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @itsbilal)

___
*[`pkg/storage/pebble.go` line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120784#gh-1532908923:-NtWRk452BKbG1T0yqZ-:b4ijjdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/63fe7580bcc0b362ef4e15acc651574cc5b00e51/pkg/storage/pebble.go#L144)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

maybe settings.SystemOnly is safer? but also good with this
</blockquote></details>

I originally had `SystemOnly`, but we build sstables within tenants for the purpose of `AddSSTable`, so it needs to be visible to the tenant when the tenant calls `MakeIngestionWriterOptions`. I added a comment describing this.
___
*[`pkg/storage/pebble.go` line 164 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120784#gh-1532920078:-NtWRjsV98IksjRDGgk2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/63fe7580bcc0b362ef4e15acc651574cc5b00e51/pkg/storage/pebble.go#L164)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

you do have a ctx in scope in `newPebble` where you call  this if you wanted it.
</blockquote></details>

Done.
___
*[`pkg/storage/sst_writer.go` line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120784#-NtWRYqyDC-CkOIS54aR:-NtWRYqyDC-CkOIS54aS:bscnqze) ([raw file](https://github.com/cockroachdb/cockroach/blob/62e54c5aa2eb5087dcf028f575962f5c4ca0b5c2/pkg/storage/sst_writer.go#L121)):*
> ```Go
> 	// reduce compression ratio.
> 	opts.BlockSize = 128 << 10
> 	opts.MergerName = "nullptr"
> ```

@dt @stevendanna — any reason not to apply this cluster setting to the backup SSTs too?

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120784)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @itsbilal, and @stevendanna)

___
*[`pkg/storage/sst_writer.go` line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120784#-NtWRYqyDC-CkOIS54aR:-NtWUCFp37jadNVkYc2v:bfxcpx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/62e54c5aa2eb5087dcf028f575962f5c4ca0b5c2/pkg/storage/sst_writer.go#L121)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

@dt @stevendanna — any reason not to apply this cluster setting to the backup SSTs too?
</blockquote></details>

We should only do that if the cluster version is 24.1 - if we're in a mixed cluster, it's possible that the backup would be restored on a node running 23.2 code right?

**Comment by dt** [comment]
No concerns about using it in backups.

It is already gated on 24.2 here.

**Review by jbowens — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120784)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @itsbilal, @RaduBerinde, and @stevendanna)

___
*[`pkg/storage/sst_writer.go` line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120784#-NtWRYqyDC-CkOIS54aR:-NtWUsrt8kXsmkz0vjzO:b-ot1mpi) ([raw file](https://github.com/cockroachdb/cockroach/blob/62e54c5aa2eb5087dcf028f575962f5c4ca0b5c2/pkg/storage/sst_writer.go#L121)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We should only do that if the cluster version is 24.1 - if we're in a mixed cluster, it's possible that the backup would be restored on a node running 23.2 code right?
</blockquote></details>

Yeah, it's already gated on the 24.1 cluster version for all usages because earlier versions of Pebble could corrupt memory when decompressing zstd (if the zstd library re-allocated rather than filling the block cache value we passed it).

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120784#-:-NtWZ3puAGaT6UDYrZr3:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt, @itsbilal, and @jbowens)

___
*[`pkg/storage/pebble.go` line 165 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120784#-NtWYZLX5vchF9rLqrQF:-NtWYZLX5vchF9rLqrQG:bpjigrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75353ab1cb85134fae34371afbee5dd25b5cbe9/pkg/storage/pebble.go#L165)):*
> ```Go
> 	case compressionAlgorithmZstd:
> 		// Pre-24.1 Pebble's implementation of zstd had bugs that could cause
> 		// in-memory corruption. We require that the 24.1 version be finalized
> ```

"24.1 version be finalized" would suggest V24_1 not V24_1Start. I'd rephrase to "We require that the cluster version is 24.1 or an upgrade version towards 24.1 which implies that all nodes are running 24.1 code".

**Review by jbowens — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120784)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @itsbilal, and @RaduBerinde)

___
*[`pkg/storage/pebble.go` line 165 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120784#-NtWYZLX5vchF9rLqrQF:-NtWflWm2F8isC6iHSJp:bs7wks4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75353ab1cb85134fae34371afbee5dd25b5cbe9/pkg/storage/pebble.go#L165)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

"24.1 version be finalized" would suggest V24_1 not V24_1Start. I'd rephrase to "We require that the cluster version is 24.1 or an upgrade version towards 24.1 which implies that all nodes are running 24.1 code".
</blockquote></details>

Done with a tweak to clarify that we're relying on the invariant that we will never run < 24.1 code again.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120784)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @itsbilal, and @msbutler)

___
*[`pkg/storage/pebble.go` line 165 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120784#-NtWYZLX5vchF9rLqrQF:-NtWoRc--q57u7NF-zAD:b-57cn97) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75353ab1cb85134fae34371afbee5dd25b5cbe9/pkg/storage/pebble.go#L165)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Done with a tweak to clarify that we're relying on the invariant that we will never run < 24.1 code again.
</blockquote></details>

:thumbsup: :thumbsup:

File: pkg/storage/pebble.go [hunk 3]
```diff
@@ -132,6 +132,50 @@ var IngestAsFlushable = settings.RegisterBoolSetting(
...
+func getCompressionAlgorithm(ctx context.Context, settings *cluster.Settings) pebble.Compression {
+	switch compressionAlgorithm.Get(&settings.SV) {
+	case compressionAlgorithmSnappy:
+		return pebble.SnappyCompression
+	case compressionAlgorithmZstd:
+		// Pre-24.1 Pebble's implementation of zstd had bugs that could cause
+		// in-memory corruption. We require that the cluster version is 24.1 or an
+		// upgrade version towards 24.1 which implies that all nodes are running
+		// 24.1 code and will never run code < 24.1 again.
+		if settings.Version.ActiveVersionOrEmpty(ctx).IsActive(clusterversion.V24_1Start) {
```
Comment (dt) [inline pkg/storage/pebble.go:170]: Actually, maybe just use `V24_1` here? unless you explicitly need a specific mid-upgrade state, we prefer gating on the final version to make it easier to answer if something is behaving like 24.1 or not, rather than a large number of subtly different answers depending on the upgrade progress

**Review by dt — APPROVED** [review]

**Comment by jbowens** [comment]
TFTRs!

bors r+
